x-pack/plugins/cross_cluster_replication/public/app/store/actions/auto_follow_pattern.js

```diff
@@ -7,14 +7,21 @@
 import { SECTIONS, API_STATUS } from '../../constants';
 import {
   loadAutoFollowPatterns as loadAutoFollowPatternsRequest,
-  createAutoFollowPattern as createAutoFollowPatternRequest,
+  getAutoFollowPattern as getAutoFollowPatternRequest,
+  saveAutoFollowPattern as saveAutoFollowPatternRequest,
 } from '../../services/api';
 import routing from '../../services/routing';
 import * as t from '../action_types';
 import { apiAction } from './api';
 
 const { AUTO_FOLLOW_PATTERN: scope } = SECTIONS;
 
+export const selectAutoFollowPattern = (name) => ({
+  type: t.AUTO_FOLLOW_PATTERN_SELECT,
+  payload: name
+});
+
+
```

Comment:
I just dug into the action-middleware-reducer pattern you created for deduplicating the API request status logic. I think the end result is terrific because it reducers the amount of duplicate logic we have, and establishes a consistent pattern we can use everywhere. I did find that two parts of the existing solution made the code difficult for me to follow:

1. It uses middleware, which is a powerful tool but one which I would prefer to avoid if possible, since it adds an additional layer of complexity.
2. The `apiAction` interface (and the interfaces of similar helper methods) accepts a configuration object and returns another configuration object, which results in a call-site that doesn't give the casual reader much information. For example, if I came to this file to try to understand what these actions did, I'd have to also understand how `apiAction` words, and `api_middleware` works, and how the reducers work, etc. In other words, the helper methods provide indirection without abstraction.

I have an idea, borrowed from the Cloud team, which I think could help address these points.

What do you think of an interface like this? I think it's a little more transparent to someone reading it. You can guess what's going on by reading the code, but the repetitive async state logic is encapsulated. I think another big advantage of this approach is that the consumer has complete control over how errors and successful responses are handled. You can show a toast, show a fatal error, whatever you want (this kind of custom error handling happened a lot in rollups: https://github.com/elastic/kibana/blob/master/x-pack/plugins/rollup/public/crud_app/store/actions/create_job.js#L38).

```js
export const loadAutoFollowPatterns = () => {
  return createApiRequest({
    requestContext: LOAD_AUTO_FOLLOW_PATTERNS,
    sendRequest: async () => {
      await loadAutoFollowPatternsRequest();
    };
    onSuccess: (dispatch, result) => {
      // We could show a toast here too, if that makes sense.
      dispatch({
        type: LOAD_AUTO_FOLLOW_PATTERNS_SUCCESS,
        payload: { result },
      };
    },
    onError: (dispatch, error) => {
      // Show a toast or a fatal error.
    },
  });
```

`createApiRequest` would look something like this. You don't even need to use middleware. 

```js
function createApiRequest({
  requestContext,
  sendRequest,
  onSuccess,
  onError,
}) {
  return async (dispatch) => {
    dispatch({
      type: API_REQUEST_START,
      payload: {
        requestContext,
      },
    });

    let result;
    try {
      result = await sendRequest();
    } catch (error) {
      dispatch({
        type: API_REQUEST_ERROR,
        payload: {
          requestContext,
          error,
        }
      });

      onError && onError(dispatch, error);

      return showApiError(error, i18n.translate('xpack.remoteClusters.loadAction.errorTitle', {
        defaultMessage: 'Error loading remote clusters',
      }));
    }

    dispatch({
      type: API_REQUEST_SUCCESS,
      payload: {
        requestContext,
      }
    });

    onSuccess && onSuccess(dispatch, result);
  });
}
```